### PR TITLE
Refactoring for Intelligent Volume Provisioning

### PR DESCRIPTION
**Pull request by aravindavk**
Added more device utils and `fstab` handling library

Updates: #466
Signed-off-by: Aravinda VK <avishwan@redhat.com>

### CODE REVIEW

File: glusterd2/volume/volume-utils.go [hunk 1]
```diff
@@ -17,6 +17,21 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+// GetRedundancy calculates redundancy count based on disperse count
+func GetRedundancy(disperse uint) int {
+	var temp, l, mask uint
+	temp = disperse
+	l = 0
```
Comment (Madhu-1) [inline glusterd2/volume/volume-utils.go]: no need to set it explicitly to zero

Comment (aravindavk) [reply glusterd2/volume/volume-utils.go]: This func is moved from `volume-create-disperse.go` to utils. Implementation not changed

**Comment by aravindavk** [comment]
Note for reviewers:

- Moved `glusterd2/commands/volumes/events.go` to `glusterd2/volume/events.go`
- Moved `getRedundancy` function from `glusterd2/commands/volumes/volume-create-disperse.go` to `glusterd2/volume/volume-utils.go` file
- Moved  `glusterd2/commands/volumes/brickutils.go` to `pkg/api/brickutils.go`
- Added three additional fields to Device Resp

         + VgName        string `json:"vg-name"`
         + AvailableSize uint64 `json:"available-size"`
         + Used          bool   `json:"used"`
- New file added for `fstab` functionalities ` plugins/device/deviceutils/fstab.go`
- More functions added to ` plugins/device/deviceutils/utils.go` 
- Updated `AvailableSize` during device add

File: glusterd2/volume/volume-utils.go (hunk 1 above)
Comment (aravindavk) [reply glusterd2/volume/volume-utils.go]: fixed.

File: pkg/api/brickutils.go [hunk 2]
```diff
@@ -1,14 +1,13 @@
...
 	"fmt"
 
-	"github.com/gluster/glusterd2/pkg/api"
-
 	"github.com/pborman/uuid"
 )
 
-func nodesFromVolumeCreateReq(req *api.VolCreateReq) ([]uuid.UUID, error) {
+// NodesFromVolumeCreateReq extracts the list of nodes from Volume Create request
+func NodesFromVolumeCreateReq(req *VolCreateReq) ([]uuid.UUID, error) {
```
Comment (prashanthpai) [inline pkg/api/brickutils.go]: You can make this as a method of VolCreateReq structure and then use it as `req.Nodes()`

File: pkg/api/brickutils.go [hunk 3]
```diff
@@ -26,7 +25,8 @@ func nodesFromVolumeCreateReq(req *api.VolCreateReq) ([]uuid.UUID, error) {
 	return nodes, nil
 }
 
-func nodesFromVolumeExpandReq(req *api.VolExpandReq) ([]uuid.UUID, error) {
+// NodesFromVolumeExpandReq extracts list of Peer IDs from Volume Expand request
+func NodesFromVolumeExpandReq(req *VolExpandReq) ([]uuid.UUID, error) {
```
Comment (prashanthpai) [inline pkg/api/brickutils.go]: Same here.

**Comment by aravindavk** [comment]
@prashanthpai @Madhu-1 addressed review comments, please review

File: plugins/device/deviceutils/fstab.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	Filename string
+	Mounts   []FstabMount
+}
+
+func lock(filename string) (*os.File, error) {
+	lockfile, err := os.Create(filename + ".lock")
+	if err != nil {
+		return nil, err
+	}
+	defer lockfile.Close()
```
Comment (prashanthpai) [inline plugins/device/deviceutils/fstab.go]: The file is closed when the function returns. Doesn't closing file description implicitly unlock ?

Comment (aravindavk) [reply plugins/device/deviceutils/fstab.go]: yes, will remove this

**Comment by prashanthpai** [comment]
retest this please

**Comment by prashanthpai** [comment]
retest this please

**Pull request merged**